File: scripts/gemini-ask.js
#!/usr/bin/env node
/**
 * gemini-ask.js - CLI tool for Google Gemini API
 * 
 * Usage:
 *   node gemini-ask.js "Your question here"
 *   echo "Your question" | node gemini-ask.js
 *   node gemini-ask.js --file prompt.txt
 *   node gemini-ask.js --model gemini-pro "Your question"
 * 
 * Environment:
 *   GEMINI_API_KEY - Required API key
 *   GEMINI_MODEL   - Optional default model (default: gemini-pro)
 */

import { GoogleGenerativeAI } from '@google/generative-ai';
import fs from 'fs';

const DEFAULT_MODEL = process.env.GEMINI_MODEL || 'gemini-pro';

async function main() {
  // Check API key
  if (!process.env.GEMINI_API_KEY) {
    console.error('Error: GEMINI_API_KEY environment variable is required');
    process.exit(1);
  }

  // Parse arguments
  let prompt = '';
  let modelName = DEFAULT_MODEL;
  const args = process.argv.slice(2);
  
  // Check for --model flag
  const modelIndex = args.findIndex(a => a === '--model' || a === '-m');
  if (modelIndex !== -1 && args[modelIndex + 1]) {
    modelName = args[modelIndex + 1];
    args.splice(modelIndex, 2);
  }

  // Check for --system flag (system prompt)
  let systemPrompt = '';
  const systemIndex = args.findIndex(a => a === '--system' || a === '-s');
  if (systemIndex !== -1 && args[systemIndex + 1]) {
    systemPrompt = args[systemIndex + 1];
    args.splice(systemIndex, 2);
  }

  if (args.includes('--file') || args.includes('-f')) {
    const fileIndex = args.findIndex(a => a === '--file' || a === '-f');
    const filePath = args[fileIndex + 1];
    if (!filePath) {
      console.error('Error: No file specified');
      process.exit(1);
    }
    prompt = fs.readFileSync(filePath, 'utf-8');
    // Remove --file and path from args
    args.splice(fileIndex, 2);
  } else if (args.length > 0) {
    prompt = args.join(' ');
  } else if (!process.stdin.isTTY) {
    // Read from stdin
    const chunks = [];
    for await (const chunk of process.stdin) {
      chunks.push(chunk);
    }
    prompt = Buffer.concat(chunks).toString('utf-8');
  }

  if (!prompt.trim()) {
    console.error('Error: No prompt provided');
    console.error('Usage: gemini-ask "your question"');
    console.error('       gemini-ask --model gemini-pro "your question"');
    console.error('       echo "your question" | gemini-ask');
    process.exit(1);
  }

  try {
    const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
    const model = genAI.getGenerativeModel({ model: modelName });

    // Prepare generation config
    const generationConfig = {
      maxOutputTokens: 8192,
      temperature: 0.7,
      topP: 0.95,
    };

    let result;
    
    if (systemPrompt) {
      // Use chat with system prompt
      const chat = model.startChat({
        generationConfig,
        history: [
          { role: 'user', parts: [{ text: systemPrompt }] },
          { role: 'model', parts: [{ text: 'Understood. I will follow that instruction.' }] },
        ],
      });
      result = await chat.sendMessage(prompt);
    } else {
      // Direct generation
      result = await model.generateContent({
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        generationConfig,
      });
    }

    const response = await result.response;
    const text = response.text();
    
    console.log(text);
  } catch (error) {
    console.error('Error:', error.message);
    if (error.message.includes('API key')) {
      console.error('Make sure GEMINI_API_KEY is set correctly');
    }
    process.exit(1);
  }
}

main();

File: perl5/gemini-ask.pl
#!/usr/bin/perl
# gemini-ask.js — portiert nach perl5
# Quelle: javascript, OpenClaw@gateway1:scripts/gemini-ask.js
# auch in: OpenClaw@gateway2:scripts/gemini-ask.js
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use open qw(:std :utf8);
use JSON qw(decode_json encode_json);
use LWP::UserAgent;
use HTTP::Request::Common qw(POST);
use Getopt::Long qw(GetOptions);
use File::Basename qw(basename);
use Encode qw(decode);

# gemini-ask.pl - CLI tool for Google Gemini API
# 
# Usage:
#   perl gemini-ask.pl "Your question here"
#   echo "Your question" | perl gemini-ask.pl
#   perl gemini-ask.pl --file prompt.txt
#   perl gemini-ask.pl --model gemini-pro "Your question"
# 
# Environment:
#   GEMINI_API_KEY - Required API key
#   GEMINI_MODEL   - Optional default model (default: gemini-pro)

my $DEFAULT_MODEL = $ENV{'GEMINI_MODEL'} || 'gemini-pro';

sub main {
    # Check API key
    my $api_key = $ENV{'GEMINI_API_KEY'};
    if (!$api_key) {
        print STDERR "Error: GEMINI_API_KEY environment variable is required\n";
        exit 1;
    }

    # Parse arguments
    my $prompt = '';
    my $modelName = $DEFAULT_MODEL;
    my $systemPrompt = '';
    my $file_path = '';
    my @remaining_args;

    GetOptions(
        'model|m=s'  => \$modelName,
        'system|s=s' => \$systemPrompt,
        'file|f=s'   => \$file_path,
        '<>'         => sub { push @remaining_args, $_[0] }
    ) or die "Error in command line arguments\n";

    if ($file_path) {
        open my $fh, '<:encoding(UTF-8)', $file_path
            or die "Error: Cannot read file '$file_path': $!\n";
        {
            local $/;
            $prompt = <$fh>;
        }
        close $fh;
    } elsif (@remaining_args) {
        $prompt = join(' ', @remaining_args);
    } elsif (!-t STDIN) {
        # Read from stdin
        my @chunks;
        while (my $chunk = <STDIN>) {
            push @chunks, $chunk;
        }
        $prompt = join('', @chunks);
    }

    if (!$prompt || !length($prompt)) {
        print STDERR "Error: No prompt provided\n";
        print STDERR "Usage: gemini-ask \"your question\"\n";
        print STDERR "       gemini-ask --model gemini-pro \"your question\"\n";
        print STDERR "       echo \"your question\" | gemini-ask\n";
        exit 1;
    }

    eval {
        # Prepare generation config
        my $generationConfig = {
            maxOutputTokens => 8192,
            temperature     => 0.7,
            topP           => 0.95,
        };

        my $ua = LWP::UserAgent->new;
        $ua->timeout(60);

        my $url;
        my $req;
        my $response;

        if ($systemPrompt) {
            # Use chat with system prompt
            $url = "https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent?key=$api_key";
            
            my $data = {
                contents => [
                    { role => 'user', parts => [{ text => $systemPrompt }] },
                    { role => 'model', parts => [{ text => 'Understood. I will follow that instruction.' }] },
                    { role => 'user', parts => [{ text => $prompt }] }
                ],
                generationConfig => $generationConfig
            };
            
            $req = POST $url,
                'Content-Type' => 'application/json',
                'Content'      => encode_json($data);
                
            $response = $ua->request($req);
        } else {
            # Direct generation
            $url = "https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent?key=$api_key";
            
            my $data = {
                contents => [
                    { role => 'user', parts => [{ text => $prompt }] }
                ],
                generationConfig => $generationConfig
            };
            
            $req = POST $url,
                'Content-Type' => 'application/json',
                'Content'      => encode_json($data);
                
            $response = $ua->request($req);
        }

        if ($response->is_success) {
            my $json_response = decode_json($response->decoded_content);
            if (exists $json_response->{candidates} && @{$json_response->{candidates}} > 0) {
                my $text = $json_response->{candidates}[0]->{content}->{parts}[0]->{text};
                print "$text\n";
            } else {
                die "No content returned from API";
            }
        } else {
            my $status = $response->code;
            my $message = $response->decoded_content;
            die "HTTP $status: $message";
        }
    };
    
    if ($@) {
        my $error = $@;
        chomp $error;
        print STDERR "Error: $error\n";
        if ($error =~ /API key/) {
            print STDERR "Make sure GEMINI_API_KEY is set correctly\n";
        }
        exit 1;
    }
}

main();
